base/multidimensional.jl

```diff
@@ -523,19 +523,23 @@ end
         L.mask[idx] && return (idx, s)
     end
 end
-# When wrapping a BitArray, lean heavily upon its internals -- this is a common
-# case. Just use the Int index and count as its state.
-@inline function iterate(L::LogicalIndex{Int,<:BitArray}, s=(0,1))
-    s[2] > length(L) && return nothing
-    i, n = s
+# When wrapping a BitArray, lean heavily upon its internals.
+@inline function iterate(L::Base.LogicalIndex{Int,<:BitArray})
+    L.sum == 0 && return nothing
     Bc = L.mask.chunks
-    while true
-        if Bc[_div64(i)+1] & (UInt64(1)<<_mod64(i)) != 0
-            i += 1
-            return (i, (i, n+1))
-        end
-        i += 1
+    return iterate(L::Base.LogicalIndex{Int,<:BitArray}, (1, @inbounds Bc[1]))
+end
+@propagate_inbounds function iterate(L::Base.LogicalIndex{Int,<:BitArray}, s)
+    Bc = L.mask.chunks
+    i1, c = s
+    while c==0
+        i1 == length(Bc) && return nothing
```

Comment:
I am not entirely sure about the desired safety here. If the iteration interface is used correctly, this will always be be inbounds. If we get passed an OOB-state, do we need to raise a correct error, are we allowed to return garbage (as long as no OOB read occurs) or are we allowed to perform an OOB read (possibly with segfault) and then return garbage?

The TEST instead of CMP was actually cribbed from the iteration code for vectors. Benchmarks and Agner Fog agree that there is no price difference between these instructions. I could change to `i1 >= length(Bc)`, with the disadvantage that misuse silently returns `nothing` instead of throwing an OOB error in `julia --check-bounds=yes`, and the advantage that we won't read OOB on most misuses, even in `julia --check-bounds=no`. I think we should prefer the OOB error?

For some reason the `UInt` version is much slower. I tried the following, with no success:
```
@inline function Base.iterate(L::Base.LogicalIndex{Int,<:BitArray})
    L.sum == 0 && return nothing
    Bc = L.mask.chunks
    return Base.iterate(L::Base.LogicalIndex{Int,<:BitArray}, (UInt(0), @inbounds Bc[1]))
end


@inline function Base.iterate(L::Base.LogicalIndex{Int,<:BitArray}, s::Tuple{UInt,UInt64})
    Bc = L.mask.chunks
    i1, c = s
    while c==0 
        i1 += 1
        i1 < length(Bc) || return nothing        
        @inbounds c = Bc[i1+1] 
    end
    tz = trailing_zeros(c)+1
    c = _BLSR(c)
    return (i1<<6 +tz, (i1, c))
end
```

Reply:
> with the disadvantage that misuse silently returns nothing instead of throwing an OOB error

```
julia> iterate([1], 2)

julia> iterate((1,), 2)
```

totally fine to return `nothing`.

Reply:
Ok, I got a fast UInt version now. Only question is about what is our preferred error behavior: `nothing` vs OOB error with `check-bounds` and OOB read with possible segfault when given an `@inbounds` invalid state. I still prefer the OOB error, tbh. Do we have a policy for that kind of question?

```
@inline _BLSR(x)= x & (x-1) 
@inline function Base.iterate(L::Base.LogicalIndex{Int,<:BitArray})
    L.sum == 0 && return nothing
    Bc = L.mask.chunks
    return Base.iterate(L::Base.LogicalIndex{Int,<:BitArray}, (1, @inbounds Bc[1]))
end

@inline function Base.iterate(L::Base.LogicalIndex{Int,<:BitArray}, s)
    Bc = L.mask.chunks
    i1, c = s
    while c==0 
        i1%UInt >= length(Bc)%UInt && return nothing
        i1 += 1
        @inbounds c = Bc[i1] 
    end
    tz = trailing_zeros(c)+1
    c = _BLSR(c)
    return ((i1-1)<<6 +tz, (i1, c))
end
```

Reply:
"When in doubt, do as the `Array`s".

So, couldn't just something like

https://github.com/JuliaLang/julia/blob/0e023d0d8255d33ba68e2fb968678d5f8c8e5276/base/array.jl#L707

be used?

Reply:
Arrays currently give nothing for OOB states, and ranges silently return garbage:
```
julia> iterate(1:10, 17)
(18, 18)
julia> iterate(collect(1:10), 17) === nothing
true
```
In view of that, I guess it would be consistent enough to return `nothing` and reap the speedup when there is no `@inbounds` annotation. So I'll change that.